Introduce state persistence for Ballerina streams

grainier opened the pull request:
## Purpose
This PR will introduce state persistence for Ballerina streams so that it will allow persisting the in-memory state of Ballerina streams (windows, aggregators, etc..) and recover in case of a failure.

## Approach
This introduces an abstract Snapshotable object with saveState and restoreState methods, and any other object (windows and aggregators) which refers (implements) that will get snapshotted periodically, allowing it to recover in case of a failure.

## Documentation
This feature can be enabled and customized by using the following config params;
 - "b7a.streaming.persistence.enabled" default = false
 - "b7a.streaming.persistence.directory" default = "snapshots"
 - "b7a.streaming.persistence.interval" default = 30 seconds

## Automation tests
 - Unit tests provided.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

hasithaa commented on stdlib/streams/pom.xml:
```diff
@@ -70,6 +70,26 @@
             <groupId>org.ballerinalang</groupId>
             <artifactId>ballerina-crypto</artifactId>
         </dependency>
+        <dependency>
```
Add these dependencies to .gradle file as well.

grainier replied:
Added in d666ea1

vinok88 commented on stdlib/streams/src/main/ballerina/streams/utils.bal:
```diff
@@ -32,3 +32,48 @@ public function createResetStreamEvent(StreamEvent event) returns StreamEvent {
+public function toStreamEvent(SnapshottableStreamEvent event) returns StreamEvent {
+    StreamEvent se = new(event.data, event.eventType, event.timestamp);
+    return se;
+}
```
Add new line.

grainier replied:
Fixed in c14babf

NatashaWso2 commented on stdlib/streams/src/test/resources/test-src/streamingv2-state-persistence-test.bal:
```diff
@@ -0,0 +1,97 @@
+function addToGlobalEmployeeArray(TeacherOutput e) {
+    globalTeacherOutputArray[index] = e;
+    index = index + 1;
+}
```
Shall we add a new line at the end of the file?

grainier replied:
fixed in 05e3985

NatashaWso2 commented on stdlib/streams/src/test/java/org/ballerinalang/stdlib/streams/BallerinaStreamsV2StatePersistenceTest.java:
```diff
@@ -0,0 +1,108 @@
+ * This contains methods to test state persistence capability in
+ * windows and aggregators of Ballerina Streaming V2.
+ *
+ * @since 0.990.3
```
```suggestion
 * @since 0.990.5
```

grainier replied:
fixed in 774848d

NatashaWso2 commented on stdlib/streams/src/main/java/org/ballerinalang/nativeimpl/Serialize.java:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * ballerina.model.streams:serialize(map obj).
+ */
```
Since this is a newly added file shall we add the version?

grainier replied:
fixed in 774848d

hasithaa commented:
@grainier Shall we incorporate code review suggestions, such as updating documentation. Also, mark PR with WIP.

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/aggregators.bal:
```diff
@@ -66,6 +67,24 @@ public type Sum object {
+    }
+
+    public function restoreState(map<any> state) {
+        any? iSum = state["iSum"];
```
Since `()` belongs to `any`
```suggestion
        any iSum = state["iSum"];
```

Applies to other similar usage too.

grainier replied:
fixed in ffc21df

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/collections.bal:
```diff
@@ -413,4 +413,87 @@ public type LinkedList object {
+};
+
+public type IntSort object {
+    public function __init() {
```
Do we need this empty `__init` method?

grainier replied:
fixed in 8d11727

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,291 @@
+import ballerina/task;
+import ballerina/config;
+import ballerina/internal;
+import ballerina/time;
```
Shall we order the imports alphabetically?

grainier replied:
fixed in 9f827ee. But shouldn't this be fixed in IDE plugin to do it automatically?

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/stream-event.bal:
```diff
@@ -62,3 +62,9 @@ public type StreamEvent object {
+public type SnapshottableStreamEvent record {
+    EventType eventType = "CURRENT";
+    int timestamp = 0;
+    map<anydata> data = {};
```
Should this be a closed record instead?

grainier replied:
fixed in 81f06f6

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/windows.bal:
```diff
@@ -495,6 +552,41 @@ public type TimeBatchWindow object {
+        SnapshottableStreamEvent?[] currentEventsList = toSnapshottableEvents(self.currentEventQueue.asArray());
+        StreamEvent? resetStreamEvt = self.resetEvent;
+        SnapshottableStreamEvent? resetEvt = (resetStreamEvt is StreamEvent)
+        ? toSnapshottableEvent(resetStreamEvt) : ();
```
Should we indent this line?

grainier replied:
fixed in c0de2de

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,291 @@
+
+// Function to initialize snapshotting.
+public function initPersistence() {
+    boolean enabled = config:getAsBoolean("b7a.streaming.persistence.enabled", default = false);
```
Since `false` is the default we can just say 
```suggestion
    boolean enabled = config:getAsBoolean("b7a.streaming.persistence.enabled");
```
right?

grainier replied:
fixed in 4858833

grainier commented on bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/serializer/providers/bvalue/BTypeBValueProviders.java:
```diff
@@ -265,4 +424,61 @@ public BRecordType toObject(BPacket packet, BValueDeserializer bValueDeserialize
+
+            try {
+                BUnionType unionType = new BUnionType();
+                FieldUtils.writeField(unionType, "typeName", typeName, true);
```
fixed in f2cb496

grainier commented on bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/serializer/providers/bvalue/LinkedHashSetBValueProvider.java:
```diff
@@ -0,0 +1,74 @@
+ * Class implements @{@link SerializationBValueProvider} to provide the mapping between {@link LinkedHashSet} and
+ * {@link BValue} representation of it.
+ *
+ * @since 0.990.3
```
fixed in 774848d

grainier commented:
> @grainier Shall we incorporate code review suggestions, such as updating documentation. Also, mark PR with WIP.

Fixed:
    - Fixed according to PR comments.
    - Fixed according to code review comments. (use Ballerina doc comments, double check persistence initialization when there are two or more forever blocks, etc...)

Possible Improvements:
    - Check the feasibility of making snapshot persistence a service. 

@hasithaa @gimantha @mohanvive If there's no changes required, shall we get this merged?

hasithaa commented on bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/serializer/JsonSerializer.java:
```diff
@@ -65,10 +66,15 @@
         bValueProvider.register(new BTypeBValueProviders.BArrayTypeBValueProvider());
+        bValueProvider.register(new BTypeBValueProviders.BMapTypeBValueProvider());
+        bValueProvider.register(new BTypeBValueProviders.BUnionTypeBValueProvider());
+        bValueProvider.register(new BTypeBValueProviders.BAttachedFunctionBValueProvider());
```
We need to clean up unnecessary logic from here.  Please create an issue, we can do it in the next iteration.

grainier replied:
Removed redundant `BUnionTypeBValueProvider` and `BAttachedFunctionBValueProvider` in 
82d9e98

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,350 @@
+# + c - Character channel to be closed.
+function closeCharChannel(any c) {
+    if (c is io:ReadableCharacterChannel) {
+        var err = c.close();
```
shouldn't we do something with `err`? same with L129?

grainier replied:
Fixed in abd66be

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,350 @@
+            return;
+        } else {
+            error e = error("Error while writing streaming state to a file.");
+            return e;
```
Should we return errors inline instead
```suggestion
            return error("Error while writing streaming state to a file.");
```

grainier replied:
Sorry, can't do. Language limitation.

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,350 @@
+
+# Function to log a persitance error.
+# + error - The `error` occured while persisting the state.
+function persitanceError(error e) {
```
`persitance` -> `persistence`? Applicable elsewhere too.

grainier replied:
Fixed in d345721

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,350 @@
+# Function to restore state of a given object.
+# + k - An unique `string` identifier for the snapshotable reference.
+# + o - The snapshotable reference to be restored.
+public function restoreState(string k, any o) {
```
Can we improve the parameter names? Since these are public functions too?

grainier replied:
Fixed in c37733b

MaryamZi commented on stdlib/streams/src/main/ballerina/streams/snapshot.bal:
```diff
@@ -0,0 +1,350 @@
+        restoreStates();
+        startPersisting();
+    }
+}
```
needs a new line.

grainier replied:
Fixed in e215d7a
